torch/distributed/elastic/rendezvous/c10d_rendezvous_backend.py

```diff
@@ -77,22 +77,37 @@ def get_state(self) -> Optional[Tuple[bytes, Token]]:
 
     def set_state(
         self, state: bytes, token: Optional[Token] = None
-    ) -> Optional[Tuple[bytes, Token]]:
+    ) -> Optional[Tuple[bytes, Token, bool]]:
         """See base class."""
         base64_state_str: str = codecs.encode(state, "base64").decode()
 
         if token:
             # Shortcut if we know for sure that the token is not valid.
             if not isinstance(token, bytes):
-                return self.get_state()
+                result = self.get_state()
+                if result is not None:
+                    tmp = *result, False
+                    # Python 3.6 does not support tuple unpacking in return
+                    # statements.
+                    return tmp
+                return None
 
             token = token.decode()
         else:
             token = self._NULL_SENTINEL
 
         base64_state: bytes = self._call_store("compare_set", self._key, token, base64_state_str)
 
-        return self._decode_state(base64_state)
+        state_token_pair = self._decode_state(base64_state)
+        if state_token_pair is None:
+            return None
+
+        new_state, new_token = state_token_pair
+
+        # C10d Store's compare_set method does not offer an easy way to find out
```

Comment:
it is possible to extend C10d store with this functionality in the long term?

Reply:
I think it would overload the `Store` type even more. Its API, as it is, is already pretty convoluted. I would avoid adding any new features to `Store` at this point.